File: dot-net/hcx-integrator-sdk/Io.HcxProtocol/Dto/BaseRequest.cs
using Io.HcxProtocol.Exceptions;
using Io.HcxProtocol.Impl;
using Io.HcxProtocol.Interfaces;
using Io.HcxProtocol.Utils;
using System;
using System.Collections.Generic;
using System.Linq;

namespace Io.HcxProtocol.Dto
{
    /**
     * Library  : Io.Hcx.Protocol
     * Author   : WalkingTree Technologies
     * Date     : 15-Mar-2023
     * All Rights Reserved. WalkingTree Technologies.
     **/

    /// <summary>
    /// This is base class to extract the protocol headers from the jwe/json payload and Validate Headers Data.
    /// </summary>
    public class BaseRequest : IBaseRequest, IDisposable
    {
        private Dictionary<string, object> Payload { get; set; }
        public Dictionary<string, object> ProtocolHeaders { get; set; }

      private static NLog.Logger _logger = NLog.LogManager.GetCurrentClassLogger();
        public BaseRequest(Dictionary<string, object> payload)
        {
            this.Payload = payload;
            if (payload.ContainsKey(Constants.PAYLOAD))
                this.ProtocolHeaders = JSONUtils.DecodeBase64String<Dictionary<string, object>>(GetPayloadValues()[0]);
            else
                this.ProtocolHeaders = payload;
        }

        public Dictionary<string, object> GetPayload()
        {
            return Payload;
        }
        public string GetWorkflowId()
        {
            return GetHeader(Constants.WORKFLOW_ID);
        }
        public string GetApiCallId()
        {
            return GetHeader(Constants.HCX_API_CALL_ID);
        }
        public string GetCorrelationId()
        {
            return GetHeader(Constants.HCX_CORRELATION_ID);
        }
        public string GetHcxSenderCode()
        {
            return GetHeader(Constants.HCX_SENDER_CODE);
        }
        public string GetTimestamp()
        {
            return GetHeader(Constants.HCX_TIMESTAMP);
        }
        public string GetDebugFlag()
        {
            return GetHeader(Constants.DEBUG_FLAG);
        }
        public string GetStatus()
        {
            return GetHeader(Constants.STATUS);
        }
        protected string GetHeader(string key)
        {
            string headerValue;

            if (key == Constants.HCX_TIMESTAMP)
            {
                try
                {
                    headerValue = Convert.ToDateTime(ProtocolHeaders[key]).ToString();
                }
                catch (Exception)
                {
                    headerValue = "";
                }
            }
            else
            {
                headerValue = ProtocolHeaders.TryGetValue(key, out var value) ? (string)value : "";
            }

            return headerValue;
        }
        protected Dictionary<string, object> GetHeaderObj(string key)
        {
            string objString = JSONUtils.Serialize(ProtocolHeaders[key]);
            return JSONUtils.Deserialize<Dictionary<string, object>>(objString);
        }

        private void SetHeaderObj(string key, object value)
        {
            ProtocolHeaders.Add(key, value);
        }

        public Dictionary<string, object> GetErrorDetails()
        {
            return GetHeaderObj(Constants.ERROR_DETAILS);
        }
        public Dictionary<string, object> GetDebugDetails()
        {
            return GetHeaderObj(Constants.DEBUG_DETAILS);
        }
        public string GetRedirectTo()
        {
            return GetHeader(Constants.REDIRECT_TO);
        }
        public string[] GetPayloadValues()
        {
            return Payload[Constants.PAYLOAD].ToString().Split('.');
        }
        public bool ValidateHeadersData(List<string> mandatoryHeaders, Operations operation, Dictionary<string, object> error)
        {
            List<string> missingHeaders = mandatoryHeaders.Where(key => !ProtocolHeaders.ContainsKey(key)).ToList();
            if (missingHeaders.Any())
            {
                error.Add(ErrorCodes.ERR_MANDATORY_HEADER_MISSING.ToString(), string.Format(ResponseMessage.INVALID_MANDATORY_ERR_MSG, string.Join(", ", missingHeaders)));
                return true;
            }
            // Validate Header values
           
            if (ValidateCondition(GetApiCallId().GetType()!= typeof(string) , error, ErrorCodes.ERR_INVALID_API_CALL_ID.ToString(), ResponseMessage.INVALID_API_CALL_ID_ERR_MSG)) //change no 5 implemented
                return true;
            if (ValidateCondition(GetCorrelationId().GetType()!=typeof(string), error, ErrorCodes.ERR_INVALID_CORRELATION_ID.ToString(), ResponseMessage.INVALID_CORRELATION_ID_ERR_MSG)) //change no 5 implemented
                return true;
            if (ValidateCondition(!GetTimestamp().IsValidTimestamp(), error, ErrorCodes.ERR_INVALID_TIMESTAMP.ToString(), ResponseMessage.INVALID_TIMESTAMP_ERR_MSG))
                return true;
            if (ValidateCondition(ProtocolHeaders.ContainsKey(Constants.WORKFLOW_ID) && GetWorkflowId().GetType()!=typeof(string), error, ErrorCodes.ERR_INVALID_WORKFLOW_ID.ToString(), ResponseMessage.INVALID_WORKFLOW_ID_ERR_MSG)) //change no 5 implemented
                return true;
            //validating optional headers
            ValidateOptionalHeaders(error);
            // validating onAction headers
            ValidateOnAction(operation, error);

            return false;
        }

        public bool ValidateJwePayload(Dictionary<string, object> error, string[] payloadArr)
        {
            bool status = false;
            try
            {
                if (payloadArr != null && payloadArr.Length != Constants.PROTOCOL_PAYLOAD_LENGTH)
                {
                    error.Add(ErrorCodes.ERR_INVALID_PAYLOAD.ToString(), ResponseMessage.INVALID_PAYLOAD_LENGTH_ERR_MSG);
                   _logger.Error(ErrorCodes.ERR_INVALID_PAYLOAD.ToString(), ResponseMessage.INVALID_PAYLOAD_LENGTH_ERR_MSG);
                    status = true;
                    return status;
                }
                if (payloadArr != null)
                {
                    foreach (string value in payloadArr)
                    {
                        if (string.IsNullOrEmpty(value))
                        {
                            error.Add(ErrorCodes.ERR_INVALID_PAYLOAD.ToString(), ResponseMessage.INVALID_PAYLOAD_VALUES_ERR_MSG);
                          _logger.Error(ErrorCodes.ERR_INVALID_PAYLOAD.ToString(), ResponseMessage.INVALID_PAYLOAD_VALUES_ERR_MSG);
                            status = true;
                            return status;
                        }
                    }
                }
                status = false;
            }
            catch(Exception ex)
            {
                error.Add("Some error occuured", ex.Message.ToString());
                _logger.Error("Some error occuured", ex.Message.ToString());

            }
            return status;
        }

        public bool ValidateCondition(bool condition, Dictionary<string, object> error, string key, string msg)
        {
            if (condition)
            {
                error.Add(key, msg);
                return true;
            }
            return false;
        }
        public bool ValidateDetails(Dictionary<string, object> inputObj, Dictionary<string, object> error, string errorKey, string msg, List<string> rangeValues, string rangeMsg)
        {
            if (inputObj == null || inputObj.Count == 0)
            {
                error.Add(errorKey, msg);
                return true;
            }
            else if (!inputObj.ContainsKey(Constants.CODE) || !inputObj.ContainsKey(Constants.MESSAGE))
            {
                error.Add(errorKey, msg);
                return true;
            }
            foreach (string key in inputObj.Keys)
            {
                if (!rangeValues.Contains(key))
                {
                    error.Add(key, rangeMsg);
                    return true;
                }
            }
            return false;
        }

        public bool ValidateOptionalHeaders(Dictionary<string, object> error)
        {
            if (ProtocolHeaders.ContainsKey(Constants.DEBUG_FLAG) && ValidateValues(GetDebugFlag(), error, ErrorCodes.ERR_INVALID_DEBUG_FLAG.ToString(), ResponseMessage.INVALID_DEBUG_FLAG_ERR_MSG, Constants.DEBUG_FLAG_VALUES, string.Format(ResponseMessage.INVALID_DEBUG_FLAG_RANGE_ERR_MSG, string.Join(", ", Constants.DEBUG_FLAG_VALUES))))
                return true;
            if (ProtocolHeaders.ContainsKey(Constants.ERROR_DETAILS))
            {
                if (ValidateDetails(GetErrorDetails(), error, ErrorCodes.ERR_INVALID_ERROR_DETAILS.ToString(), ResponseMessage.INVALID_ERROR_DETAILS_ERR_MSG, Constants.ERROR_DETAILS_VALUES, string.Format(ResponseMessage.INVALID_ERROR_DETAILS_RANGE_ERR_MSG, string.Join(", ", Constants.ERROR_DETAILS_VALUES))))
                    return true;
                if (ValidateCondition(!Constants.RECIPIENT_ERROR_VALUES.Contains(((Dictionary<string, object>)ProtocolHeaders[Constants.ERROR_DETAILS])[Constants.CODE]), error, ErrorCodes.ERR_INVALID_ERROR_DETAILS.ToString(), ResponseMessage.INVALID_ERROR_DETAILS_CODE_ERR_MSG))
                    return true;
            }
            if (ProtocolHeaders.ContainsKey(Constants.DEBUG_DETAILS))
            {
                return ValidateDetails(GetDebugDetails(), error, ErrorCodes.ERR_INVALID_DEBUG_DETAILS.ToString(), ResponseMessage.INVALID_DEBUG_DETAILS_ERR_MSG, Constants.ERROR_DETAILS_VALUES, string.Format(ResponseMessage.INVALID_DEBUG_DETAILS_RANGE_ERR_MSG, string.Join(", ", Constants.ERROR_DETAILS_VALUES)));
            }
            return false;
        }

        public bool ValidateOnAction(Operations operation, Dictionary<string, object> error)
        {
            if (operation.getOperation().Contains("on_"))
            {
                if (ValidateCondition(!ProtocolHeaders.ContainsKey(Constants.STATUS), error, ErrorCodes.ERR_INVALID_STATUS.ToString(), string.Format(ResponseMessage.INVALID_MANDATORY_ERR_MSG, Constants.STATUS)))
                    return true;
                return ValidateValues(GetStatus(), error, ErrorCodes.ERR_INVALID_STATUS.ToString(), ResponseMessage.INVALID_STATUS_ERR_MSG, Constants.RESPONSE_STATUS_VALUES, string.Format(ResponseMessage.INVALID_STATUS_ON_ACTION_RANGE_ERR_MSG, string.Join(", ", Constants.RESPONSE_STATUS_VALUES)));
            }
            else
            {
                if (ProtocolHeaders.ContainsKey(Constants.STATUS))
                {
                    return ValidateValues(GetStatus(), error, ErrorCodes.ERR_INVALID_STATUS.ToString(), ResponseMessage.INVALID_STATUS_ERR_MSG, Constants.REQUEST_STATUS_VALUES, string.Format(ResponseMessage.INVALID_STATUS_ACTION_RANGE_ERR_MSG, string.Join(", ", Constants.REQUEST_STATUS_VALUES)));
                }
            }
            return false;
        }

        public bool ValidateValues(string inputStr, Dictionary<string, object> error, string key, string msg, List<string> statusValues, string rangeMsg)
        {
            if (string.IsNullOrEmpty(inputStr))
            {
                error.Add(key, msg);
                return true;
            }
            else if (!statusValues.Contains(inputStr))
            {
                error.Add(key, rangeMsg);
                return true;
            }
            return false;
        }

        private bool disposedValue;
        protected virtual void Dispose(bool disposing)
        {
            // check if already disposed
            if (!disposedValue)
            {
                if (disposing)
                {
                    // free managed objects here
                    Payload = null;
                    ProtocolHeaders = null;
                }

                // free unmanaged objects here

                // set the bool value to true
                disposedValue = true;
            }
        }

        public void Dispose()
        {
            Dispose(disposing: true);
            GC.SuppressFinalize(this);
        }

        ~BaseRequest() { Dispose(disposing: false); }
    }
}

File: java/hcx-integrator-sdk/src/main/java/io/hcxprotocol/dto/BaseRequest.java
package io.hcxprotocol.dto;

import io.hcxprotocol.exception.ErrorCodes;
import io.hcxprotocol.utils.DateTimeUtils;
import io.hcxprotocol.utils.JSONUtils;
import io.hcxprotocol.utils.Operations;
import org.apache.commons.collections.MapUtils;
import org.apache.commons.lang3.StringUtils;

import java.text.MessageFormat;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static io.hcxprotocol.utils.Constants.*;
import static io.hcxprotocol.utils.ResponseMessage.*;

/**
 * This is base class to extract the protocol headers from the jwe/json payload.
 */
public class BaseRequest {

    public Map<String, Object> protocolHeaders;
    Map<String, Object> payload;

    public BaseRequest(Map<String, Object> payload) throws Exception {
        this.payload = payload;
        if (payload.containsKey(PAYLOAD))
            this.protocolHeaders = JSONUtils.decodeBase64String(getPayloadValues()[0], Map.class);
        else
            this.protocolHeaders = payload;
    }

    public String getWorkflowId() {
        return getHeader(WORKFLOW_ID);
    }

    public String getApiCallId() {
        return getHeader(HCX_API_CALL_ID);
    }

    public String getCorrelationId() {
        return getHeader(HCX_CORRELATION_ID);
    }

    public String getHcxSenderCode() {
        return getHeader(HCX_SENDER_CODE);
    }

    public String getTimestamp() {
        return getHeader(HCX_TIMESTAMP);
    }

    public String getDebugFlag() {
        return getHeader(DEBUG_FLAG);
    }

    public String getStatus() {
        return getHeader(STATUS);
    }

    protected String getHeader(String key) {
        return (String) protocolHeaders.getOrDefault(key, "");
    }

    public Map<String, Object> getPayload() {
        return payload;
    }

    protected Map<String, Object> getHeaderMap(String key) {
        return (Map<String, Object>) protocolHeaders.getOrDefault(key, new HashMap<>());
    }

    public Map<String, Object> getHeaders(){
        return this.protocolHeaders;
    }

    private void setHeaderMap(String key, Object value) {
        protocolHeaders.put(key, value);
    }

    public Map<String, Object> getErrorDetails() {
        return getHeaderMap(ERROR_DETAILS);
    }

    public Map<String, Object> getDebugDetails() {
        return getHeaderMap(DEBUG_DETAILS);
    }

    public String getRedirectTo() {
        return getHeader(REDIRECT_TO);
    }

    public String[] getPayloadValues() {
        return ((String) getPayload().get(PAYLOAD)).split("\\.");
    }

    public boolean validateHeadersData(List<String> mandatoryHeaders, Operations operation, Map<String, Object> error) {
        boolean result = false;
        List<String> missingHeaders = mandatoryHeaders.stream().filter(key -> !protocolHeaders.containsKey(key)).collect(Collectors.toList());
        if (!missingHeaders.isEmpty()) {
            error.put(ErrorCodes.ERR_MANDATORY_HEADER_MISSING.toString(), MessageFormat.format(INVALID_MANDATORY_ERR_MSG, missingHeaders));
            return true;
        }
        // Validate Header values
        if (validateCondition(!isString(getHeaders().get(HCX_API_CALL_ID)), error, ErrorCodes.ERR_INVALID_API_CALL_ID.toString(), INVALID_API_CALL_ID_ERR_MSG))
            return true;
        if (validateCondition(!isString(getHeaders().get(HCX_CORRELATION_ID)), error, ErrorCodes.ERR_INVALID_CORRELATION_ID.toString(), INVALID_CORRELATION_ID_ERR_MSG))
            return true;
        if (validateCondition(!DateTimeUtils.validTimestamp(getTimestamp()), error, ErrorCodes.ERR_INVALID_TIMESTAMP.toString(), INVALID_TIMESTAMP_ERR_MSG))
            return true;
        if (validateCondition(protocolHeaders.containsKey(WORKFLOW_ID) && !isString(getHeaders().get(WORKFLOW_ID)), error, ErrorCodes.ERR_INVALID_WORKFLOW_ID.toString(), INVALID_WORKFLOW_ID_ERR_MSG))
            return true;
        //validating option headers
        result = validateOptionalHeaders(error);
        // validating onAction headers
        result = validateOnAction(operation,error);
        return result;

    }
    public boolean validateJwePayload(Map<String, Object> error, String[] payloadArr) {
        if (payloadArr != null && payloadArr.length != PROTOCOL_PAYLOAD_LENGTH) {
            error.put(ErrorCodes.ERR_INVALID_PAYLOAD.toString(), INVALID_PAYLOAD_LENGTH_ERR_MSG);
            return true;
        }
        if (payloadArr != null) {
            for (String value : payloadArr) {
                if (value == null || value.isEmpty()) {
                    error.put(ErrorCodes.ERR_INVALID_PAYLOAD.toString(), INVALID_PAYLOAD_VALUES_ERR_MSG);
                    return true;
                }
            }
        }
        return false;
    }

    public boolean validateCondition(Boolean condition, Map<String, Object> error, String key, String msg) {
        if (condition) {
            error.put(key, msg);
            return true;
        }
        return false;
    }

    public boolean isString(Object value) {
        return value instanceof String && !StringUtils.isEmpty((String) value);
    }

    public boolean validateDetails(Map<String, Object> inputMap, Map<String, Object> error, String errorKey, String msg, List<String> rangeValues, String rangeMsg) {
        if (MapUtils.isEmpty(inputMap)) {
            error.put(errorKey, msg);
            return true;
        } else if (!inputMap.containsKey(CODE) || !inputMap.containsKey(MESSAGE)) {
            error.put(errorKey, msg);
            return true;
        }
        for (String key : inputMap.keySet()) {
            if (!rangeValues.contains(key)) {
                error.put(key, rangeMsg);
                return true;
            }
        }
        return false;
    }

    public boolean validateOptionalHeaders(Map<String, Object> error) {
        if (protocolHeaders.containsKey(DEBUG_FLAG) && validateValues(getDebugFlag(), error, ErrorCodes.ERR_INVALID_DEBUG_FLAG.toString(), INVALID_DEBUG_FLAG_ERR_MSG, DEBUG_FLAG_VALUES, MessageFormat.format(INVALID_DEBUG_FLAG_RANGE_ERR_MSG, DEBUG_FLAG_VALUES)))
                return true;
        if (protocolHeaders.containsKey(ERROR_DETAILS)) {
            if (validateDetails(getErrorDetails(), error, ErrorCodes.ERR_INVALID_ERROR_DETAILS.toString(), INVALID_ERROR_DETAILS_ERR_MSG, ERROR_DETAILS_VALUES, MessageFormat.format(INVALID_ERROR_DETAILS_RANGE_ERR_MSG, ERROR_DETAILS_VALUES)))
                return true;
            if (validateCondition(!RECIPIENT_ERROR_VALUES.contains(((Map<String, Object>) protocolHeaders.get(ERROR_DETAILS)).get(CODE)), error, ErrorCodes.ERR_INVALID_ERROR_DETAILS.toString(), INVALID_ERROR_DETAILS_CODE_ERR_MSG))
                return true;

        }
        if (protocolHeaders.containsKey(DEBUG_DETAILS)) {
            return validateDetails(getDebugDetails(), error, ErrorCodes.ERR_INVALID_DEBUG_DETAILS.toString(), INVALID_DEBUG_DETAILS_ERR_MSG, ERROR_DETAILS_VALUES, MessageFormat.format(INVALID_DEBUG_DETAILS_RANGE_ERR_MSG, ERROR_DETAILS_VALUES));
        }
        return false;
    }

    public boolean validateOnAction(Operations operation, Map<String,Object> error){
        if (operation.getOperation().contains("on_")) {
            if (validateCondition(!protocolHeaders.containsKey(STATUS), error, ErrorCodes.ERR_INVALID_STATUS.toString(), MessageFormat.format(INVALID_MANDATORY_ERR_MSG, STATUS)))
                return true;
            return validateValues(getStatus(), error, ErrorCodes.ERR_INVALID_STATUS.toString(), INVALID_STATUS_ERR_MSG, RESPONSE_STATUS_VALUES, MessageFormat.format(INVALID_STATUS_ON_ACTION_RANGE_ERR_MSG, RESPONSE_STATUS_VALUES));
        }
        else {
            if (protocolHeaders.containsKey(STATUS)) {
                return validateValues(getStatus(), error, ErrorCodes.ERR_INVALID_STATUS.toString(), INVALID_STATUS_ERR_MSG, REQUEST_STATUS_VALUES, MessageFormat.format(INVALID_STATUS_ACTION_RANGE_ERR_MSG, REQUEST_STATUS_VALUES));
                }
        }
        return false;
    }



    public boolean validateValues(String inputStr, Map<String, Object> error, String key, String msg, List<String> statusValues, String rangeMsg) {
        if (StringUtils.isEmpty(inputStr)) {
            error.put(key, msg);
            return true;
        } else if (!statusValues.contains(inputStr)) {
            error.put(key, rangeMsg);
            return true;
        }
        return false;
    }

}
